core/modules/qdisp/Executive.cc

```diff
@@ -353,35 +357,34 @@ void Executive::_unTrack(int jobId) {
             untracked = true;
             if (_incompleteJobs.empty()) _allJobsComplete.notify_all();
         }
-        size = _incompleteJobs.size();
-        // Log up to 5 incomplete jobs. Very useful when jobs do not finish.
-        int c = 0;
-        for(auto j = _incompleteJobs.begin(), e = _incompleteJobs.end(); j != e && c < 5; ++j) {
-            s << j->first << " ";
-            ++c;
+        if (!untracked || LOG_CHECK_LVL(_log, LOG_LVL_DEBUG)) {
+            // Log up to 5 incomplete jobs. Very useful when jobs do not finish.
+            s = _getIncompleteJobsString(5);
         }
     }
     LOGS(_log, (untracked ? LOG_LVL_DEBUG : LOG_LVL_WARN),
          "Executive UNTRACKING " << QueryIdHelper::makeIdStr(_id, jobId)
-         << " size=" << size << " " << (untracked ? "success":"failed") << "::" << s.str());
+             << " " << (untracked ? "success":"failed") << "::" << s);
 }
 
-/// Remove all jobs from the _incompleteJobs map that have errors.
-// This function only acts when there are errors. In there are no errors,
-// markCompleted() does the cleanup, while we are waiting (in _waitAllUntilEmpty()).
-void Executive::_reapRequesters(std::unique_lock<std::mutex> const&) {
-    for(auto iter=_incompleteJobs.begin(), e=_incompleteJobs.end(); iter != e;) {
-        if (!iter->second->getDescription().respHandler()->getError().isNone()) {
-            // Requester should have logged the error to the messageStore
-            LOGS(_log, LOG_LVL_DEBUG, "Executive reaped requester for "
-                 << QueryIdHelper::makeIdStr(_id, iter->first));
-            iter = _incompleteJobs.erase(iter);
-        } else {
-            ++iter;
-        }
+
+/// _incompleteJobsMutex must be held before calling this function.
+/// @return: a string containing a list of incomplete jobs containing up to 'maxToList' jobs.
+///          If maxToList is less than 0, all jobs are printed
+std::string Executive::_getIncompleteJobsString(int maxToList) {
+    std::ostringstream os;
+    int size = _incompleteJobs.size();
+    int c = 0;
+    if (maxToList < 0) maxToList = size;
```

Comment:
Since you're already holding a mutex '_incompleteJobsMutex ' preventing the list of incomplete jobs to change then this could become:

  if (maxToList <0) maxToList = _incompleteJobs.size();